pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,907 @@
+    #[cfg(feature = "runtime-benchmarks")]
+    fn bench_set_balance(asset_id: &AssetId, account: &AccountId, amount: Balance);
+}
+
+#[pallet(dev_mode)]
+pub mod pallet {
+    use super::*;
+
+    /// Pooled Staking pallet.
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// Type used to represent stream ids. Should be large enough to not overflow.
+        type StreamId: AtLeast32BitUnsigned
+            + Default
+            + Debug
+            + Copy
+            + Clone
+            + FullCodec
+            + TypeInfo
+            + MaxEncodedLen;
+
+        /// The balance type, which is also the type representing time (as this
+        /// pallet will do math with both time and balances to compute how
+        /// much should be paid).
+        type Balance: Balance;
+
+        /// Type representing an asset id, a identifier allowing distinguishing assets.
+        type AssetId: Debug + Clone + FullCodec + TypeInfo + MaxEncodedLen + PartialEq + Eq;
+
+        /// Provide interaction with assets.
+        type Assets: Assets<Self::AccountId, Self::AssetId, Self::Balance>;
+
+        /// Represents which units of time can be used. Designed to be an enum
+        /// with a variant for each kind of time source/scale supported.
+        type TimeUnit: Debug + Clone + FullCodec + TypeInfo + MaxEncodedLen + Eq;
+
+        /// Provide the current time in given unit.
+        type TimeProvider: TimeProvider<Self::TimeUnit, Self::Balance>;
+
+        type WeightInfo;
+    }
+
+    type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+    type AssetIdOf<T> = <T as Config>::AssetId;
+
+    pub type RequestNonce = u32;
+
+    /// A stream payment from source to target.
+    /// Stores the last time the stream was updated, which allows to compute
+    /// elapsed time and perform payment.
+    #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+    #[derive(RuntimeDebug, PartialEq, Eq, Encode, Decode, Clone, TypeInfo)]
+    pub struct Stream<AccountId, Unit, AssetId, Balance> {
+        /// Payer, source of the stream.
+        pub source: AccountId,
+        /// Payee, target of the stream.
+        pub target: AccountId,
+        /// Steam config (time unit, asset id, rate)
+        pub config: StreamConfig<Unit, AssetId, Balance>,
+        /// How much is deposited to fund this stream.
+        pub deposit: Balance,
+        /// Last time the stream was updated in `config.time_unit`.
+        pub last_time_updated: Balance,
+        /// Nonce for requests. This prevents a request to make a first request
+        /// then change it to another request to frontrun the other party
+        /// accepting.
+        pub request_nonce: RequestNonce,
```

Comment:
You suggest removing the nonce parameter from the `accept_requested_change` call, and preventing accepting a request if one have been proposed before in the same block? What about delays and congestions? Approval extrinsic might be delayed by some blocks, which the requester can frontrun by paying high fees and go throught in a block before. I think the nonce is a simple solution to avoid frontrunning, and the dapp can hide that easily.

Reply:
I mean using the block number as the nonce. So when the user wants to approve a request, they would need to read the proposed config which would include the block number at which it was proposed. If a new change is proposed in the same block, it will be rejected. And if a new change is proposed in a later block, it will overwrite the old one but the user will not accidentally accept it, as it will have a different block number.

The main benefit is that we don't need to explain the concept of "request nonce", it's just the block number, but the drawback is that it only allows one change per block (per stream).

Reply:
In that case it may even not be necessary to forbid changes in the same block. As long as the other party see the proposal in an already proposed block and use that block number afterward to accept it should be safe, right?

Reply:
You still need to provide the block number at which the request was made to accept it right? I am not sure it is more intuitive tbh